jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+        (psutil.Process): psutil.Process object of the current process.
     """
     return get_process_object(os.getpid())
 
 
 def get_process_object(pid, die=True):
-    """Get Process object of a process id
+    """Get psutil.Process object of a given process ID (PID).
 
-    Arguments:
-        pid (int) -- pid of the process
-
-    Keyword Arguments:
-        die {bool} -- die if process not found (default: {True})
+    Args:
+        pid (int): Process ID (PID) to get
+        die (bool, optional): Whether to raise an exception if no process with the given PID is found in the \
+            current process list or not. Defaults to True.
 
     Raises:
-        psutil.NoSuchProcess: if process not found and die = True
+        psutil.NoSuchProcess: If process with the given PID is not found and die set to True.
+        psutil.AccessDenied: If permission denied.
 
     Returns:
-        [psutil.Process] -- process object
+        psutil.Process or None: psutil.Process object if he given PID is found, None otherwise.
     """
     try:
         return psutil.Process(pid)
-    except psutil.NoSuchProcess as e:
+    except (psutil.ZombieProcess, psutil.AccessDenied, psutil.NoSuchProcess) as e:
+        # when you query processess owned by another user, especially on macOS and Windows you may get AccessDenied exception
         if die:
             raise e
         else:
             return None
 
 
 def get_user_processes(user):
-    """Get all process for a specific user
+    """Get all process for a specific user.
 
-    Arguments:
-        user {str} -- username
+    Args:
+        user (str): Te user name to match against.
 
-    Returns:
-        [list(int)] -- list of process pids for that user
+    Yields:
+        psutil.Process: psutil.Process object for all processes owned by `user`.
     """
-    result = []
-    for process in psutil.process_iter():
-        if process.username() == user:
-            result.append(process.pid)
-    return result
+    try:
+        for process in psutil.process_iter(["name", "exe", "cmdline"]):
+            if process.username() == user:
+                yield process
+    except (psutil.AccessDenied, psutil.NoSuchProcess):
+        pass
+
 
+def kill_user_processes(user, sig=signal.SIGTERM, timeout=5, sure_kill=False):
+    """Kill all processes for a specific user.
 
-def kill_user_processes(user):
-    """Kill all processes for a specific user
+    Args:
+        user (str): The user name to match against.
+        sig (signal, optional): See signal module constants. Defaults to signal.SIGTERM.
+        timeout (int, optional): How long to wait for a process to terminate (seconds) before raise exception or, if sure_kill=True, send a SIGKILL. Defaults to 5.
+        sure_kill (bool, optional): Whether to fallback to SIGKILL if the timeout exceeded for the terminate operation or not. Defaults to False.
 
-    Arguments:
-        user {str} -- username
+    Returns:
```

Comment:
We can return empty list instead `None`?

Reply:
the same as the previous replay.